fix: Create Step Input component matching Fiori3

salarenko opened the pull request:
#### Please provide a link to the associated issue.
Closes: #2386 
Closes: #1929

Sorybook section will be created in separate PR closing this ticket: #2783

#### Please provide a brief summary of this pull request.
This pull request introduces new `StepInput` component matching Fiori3 styling and behavior.

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/master/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist

Documentation checklist:
- [x] update `README.md`
- [x] [Breaking Changes Wiki](https://github.com/SAP/fundamental-ngx/wiki/Breaking-Changes)
- [x] Documentation Examples
- [x] Stackblitz works for all examples

InnaAtanasova commented on libs/core/src/lib/step-input/step-input.component.ts:
```diff
@@ -0,0 +1,475 @@
+
+            this._subscriptions.add(
+                this._setupButtonListener(this.decrementButton)
+                    .subscribe(() => {
```
Did you forget to unsubscribe?

InnaAtanasova commented:
The PR should not be named `feat`

salarenko commented on libs/core/src/lib/step-input/step-input.component.html:
```diff
@@ -0,0 +1,51 @@
+        'is-information': state === 'information',
+        'fd-step-input--without-buttons': !hasStepButtons,
+        'fd-step-input--compact': compact
+     }">
```
I don't like that syntax :/ but I will switch to `[ngClass]="state ? 'is-' + state : ''"` to be consistent with other components.

salarenko commented:
@InnaAtanasova 
I've made following changes:
- Introduced disabled and read-only state for `StepInput`
- Increment/Decrement button gets disabled based on min/max value boundary
- Now when input is focused `is-focus` class is set (as Fundamental Styles documentation says)
- Introduced incrementing/decrementing value with mouse wheel
- Introduced `Input() textAlign` to align text inside input
- Introduced unsubscribing on `onDestroy`
- Introduced `_cutFloatingNumberDistortion` function to cut distortion caused by floating point based calculations (`0.1 + 0.2 = 0.30000000000000004`) 
- Introduced new StepInput documentation class for the examples
-  Splitted code examples into separate variables to attach them properly to the examples (`step-input-docs.component.ts`)

stefanoScalzo commented on libs/core/src/lib/step-input/step-input.component.html:
```diff
@@ -0,0 +1,55 @@
+            (click)="onTouched()">
+    </button>
+</div>
+<span class="fd-form-label fd-form-label--unit-description">
```
Use ngIf so that it doesnt have padding or an empty label in case user does not add one

stefanoScalzo commented on apps/docs/src/app/core/component-docs/step-input/step-input-docs.component.html:
```diff
@@ -0,0 +1,108 @@
+        <li>Input text alignment</li>
+        <li>Optional format grouping</li>
+        <li>Setting upper and lower boundary limits for the value</li>
+        <li>Specifying minimal or maximal number of fraction digits</li>
```
*fractional

stefanoScalzo commented on apps/docs/src/app/core/component-docs/step-input/step-input-docs.component.html:
```diff
@@ -0,0 +1,108 @@
+    Step Input label
+</fd-docs-section-title>
+<description>
+    Step Input label is displayed nest to the step control and should be used to describe the unit of the value selected
```
nest or next?

fkolar commented:
@salarenko I see currency* properties inside this components, I think step input should have nothing to do with a currency, it should just increment and decrement a number or whatever source you give it to it. 

It should be generic component and user should be able to composite if they want to increment money fields, or any other number, quantites, percentages

salarenko commented:
@stefanoScalzo 
I've fixed the typos and now step input label is displayed conditionally only when its needed.

JKMarkowski commented on libs/core/src/lib/step-input/step-input.component.ts:
```diff
@@ -0,0 +1,555 @@
+    decrementButton: ElementRef;
+
+    /** @hidden */
+    @ViewChild('inputElement', {read: ElementRef, static: true})
```
You need this reference in ngOnInit?

salarenko replied:
Yes, this is because `writeValue` is called before view initialization and this method calls `_updateViewValue` which requires reference to HTML `input` element.
